File: 140.cpp
class Solution {
public:
    /*
     * @param a, b, n: 32bit integers
     * @return: An integer
     */
    int fastPower(int a, int b, int n) {
        if (n == 0) {
            return 1 % b;
        }
        if (n == 1) {
            return a % b;
        }
        long long temp = fastPower(a, b, n / 2);
        if (n % 2 == 0) {
            return (temp * temp) % b;
        } else {
            return ((temp * temp) % b * a) % b;
        }
    }
};

File: 140.py


class Solution:
    """
    @param a, b, n: 32bit integers
    @return: An integer
    """
    def fastPower(self, a, b, n):
        ans = 1
        while b > 0:
            if b % 2==1:
                ans = ans * a % n
            a = a * a % n
            b = b / 2
        return ans % n

        # WA because of lintcode 
        # AC